### PR TITLE
Mf 359 reservation function acceptance tests

### PR DESCRIPTION
**Pull request by PeterDavidWilliams**

### CODE REVIEW

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿Feature: Reservation created
+	In order to know when a reservation has been created
+	I want to be told when the reservation search index is updated
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature]: Slightly wrong wording should be something more like
`I want to be notified when the reservation is created and the search index is updated`

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/DeleteReservation.feature [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+Scenario: Reservation search index should be updated on reservation deletion
+	Given I have a reservation ready for deletion
+	When a delete reservation event is triggered
+	Then the reservation search index should be updated with the deleted reservation removed
+
+Scenario: Employer should be notified on reservation deletion
+	Given I have a reservation ready for creation
+	When a delete reservation event is triggered
+	Then the employer should be notified of the deleted reservation
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/DeleteReservation.feature]: Would be good to add another scenario - this one is actually covering when a provider has deleted an employers reservation. Would also be good to cover the following:

- Employer deletes own reservation - no email should be sent
- Employer deletes reservation by provider - no email should be sent

Also look at covering Employers with different notification settings

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+Scenario: Reservation search index should be updated on reservation creation
+	Given I have a reservation ready for creation
+	When a create reservation event is triggered
+	Then the reservation search index should be updated with the new reservation
+
+Scenario: Employer should be notified on reservation creation
+	Given I have a reservation ready for creation
+	When a create reservation event is triggered
+	Then the employer should be notified of the created reservation
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature]: This is from the Scenario Provider creates reservation on Employers behalf, would also be good to cover the following:
- Employer creates reservation - not notified
- Employer creates levy reservation - not notified

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationCreatedSteps.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        public void WhenACreateReservationEventIsTriggered()
+        {
+            var permissionRepository = Services.GetService<IProviderPermissionRepository>();
+            var mockPermissionRepository = Mock.Get(permissionRepository);
+
+            var providerPermission = new ProviderPermission
+            { AccountId = 1, AccountLegalEntityId = 1, CanCreateCohort = true, ProviderId = 1 };
+
+            mockPermissionRepository.Setup(x => x.GetAllForAccountLegalEntity(TestData.AccountLegalEntity.AccountLegalEntityId))
+                .Returns(new List<ProviderPermission> { providerPermission });
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationCreatedSteps.cs]: Not sure if its worth moving this into the TestData setup?

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationDeletedSteps.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        
+        [When(@"a delete reservation event is triggered")]
+        public void WhenADeleteReservationEventIsTriggered()
+        {
+            var accountsService = Services.GetService<IAccountsService>();
+            var mockAccountsService = Mock.Get(accountsService);
+
+            var userDetails = new UserDetails { CanReceiveNotifications = true, Email = "", Name = "", Role = "Owner", Status = 1, UserRef = "" };
+
+            mockAccountsService.Setup(x => x.GetAccountUsers(It.IsAny<long>())).ReturnsAsync(new List<UserDetails> { userDetails });
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationDeletedSteps.cs]: Duplicated in the above file, can it be moved up?

File: src/SFA.DAS.Reservations.Functions.Reservations.UnitTests/WhenReservationCreatedEventTriggered.cs [hunk 6]
```diff
@@ -18,40 +19,50 @@ public class WhenReservationCreatedEventTriggered
...
-            notifyAction.Verify(s => s.Execute(It.Is<ReservationCreatedNotificationEvent>(ev => 
+            notifyAction.Verify(s => s.Execute(It.Is<ReservationCreatedNotificationEvent>(ev =>
                 ev.Id == createdEvent.Id)), Times.Once);
         }
 
         [Test, AutoData]
         public async Task Then_Update_Index_Action_Executed(
             int courseLevel,
-            ReservationCreatedEvent createdEvent)
+            ReservationCreatedEvent createdEvent)//TODO Mock the handler and do setup on it
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Functions.Reservations.UnitTests/WhenReservationCreatedEventTriggered.cs]: Can the TODO be removed

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature (hunk 1 above)
Comment (PeterDavidWilliams) [reply src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature]: Done

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationCreatedSteps.cs (hunk 4 above)
Comment (PeterDavidWilliams) [reply src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationCreatedSteps.cs]: Done

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationDeletedSteps.cs (hunk 5 above)
Comment (PeterDavidWilliams) [reply src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Steps/ReservationDeletedSteps.cs]: Done

File: src/SFA.DAS.Reservations.Functions.Reservations.UnitTests/WhenReservationCreatedEventTriggered.cs (hunk 6 above)
Comment (PeterDavidWilliams) [reply src/SFA.DAS.Reservations.Functions.Reservations.UnitTests/WhenReservationCreatedEventTriggered.cs]: Done

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/DeleteReservation.feature (hunk 2 above)
Comment (PeterDavidWilliams) [reply src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/DeleteReservation.feature]: Done

File: src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature (hunk 3 above)
Comment (PeterDavidWilliams) [reply src/SFA.DAS.Reservations.Functions.Reservations.AcceptanceTests/Features/CreateReservation.feature]: Done

**Pull request merged**